From 8000597349d447b1dc571fd600508b8e4ec9f1ce Mon Sep 17 00:00:00 2001
From: Venkatasaiprudhvi Sannidhi <vsannidhi@maxlinear.com>
Date: Tue, 10 Jun 2025 10:58:10 +0530
Subject: [PATCH] WLANRTSYS-88849 [Wav700] [MLD] Intel BE200 is unable to
 associate with MLD in OWE security - Invalid AKMP

Issue: Association with Intel BE200 with market driver fails since the hostapd would reject the association response with the reason code "Invalid AKMP".

Root cause: In the ML IE->per STA profile for linked STA in association request, it is noticed that the non-inheritance extension element with the extension list has the Element ID of Diffie-Hellman Parameter (32).
Hence the current hostapd-ng (which is in alignemnt with the Opesource hostapd) would clear the DH params which is inherited from assoc link.

Fix: Ignore the DH Ext Tag in the non-inheritance list and thereby the associated link's DH params will be considered and verified that the association succeeds.

Unit test: Passed with the reported client - driver combination.
---
 src/common/ieee802_11_common.c | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/src/common/ieee802_11_common.c b/src/common/ieee802_11_common.c
index aa72b2e22..3a2abf264 100644
--- a/src/common/ieee802_11_common.c
+++ b/src/common/ieee802_11_common.c
@@ -1081,10 +1081,12 @@ void ieee802_11_elems_clear_ext_ids(struct ieee802_11_elems *elems,
 		case WLAN_EID_EXT_FILS_NONCE:
 			elems->fils_nonce = NULL;
 			break;
+#ifndef CONFIG_VENDOR_MXL
 		case WLAN_EID_EXT_OWE_DH_PARAM:
 			elems->owe_dh = NULL;
 			elems->owe_dh_len = 0;
 			break;
+#endif /* CONFIG_VENDOR_MXL */
 		case WLAN_EID_EXT_PASSWORD_IDENTIFIER:
 			elems->password_id = NULL;
 			elems->password_id_len = 0;
-- 
2.43.0

